boot/assets.go

```diff
@@ -0,0 +1,59 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot
+
+import (
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/gadget"
+)
+
+func InitialSealObserver(model *asserts.Model) gadget.ContentWriteObserver {
+	if model.Grade() == asserts.ModelGradeUnset {
+		// no need to observe updates when assets are not managed
+		return nil
+	}
+
+	return &initialSealObserver{
+		model: model,
+	}
+}
+
+type initialSealObserver struct {
+	model *asserts.Model
+}
+
+// Implements gadget.FileWriteObserver.
+func (o *initialSealObserver) Observe(op gadget.ObserveAction, affectedStruct *gadget.LaidOutStructure, root, realSource, relativeTarget string) (gadget.ObserveResult, error) {
+	// TODO:UC20:
+	// steps on write action:
+	// - copy new asset to assets cache
```

Comment:
I think this would require some tweaking of the secboot API. Right now we need to provide the paths to individual files, which are later hashed in the secboot library.

Reply:
PR #9084 computes a 10%+1MB safety margin for upgrades (which I presume also applies to gadget updates?), but if the cache is already populated this shouldn't be a concern.